qtfred/src/mission/dialogs/ShipEditor/ShipWeaponsDialogModel.cpp

```diff
@@ -0,0 +1,379 @@
+					delete turretBank;
+				}
+			}
+		}
+	} else {
+		for (int i = 0; i < MAX_SHIP_SECONDARY_BANKS; i++) {
+			if (SecondaryBanks[0]->getByBankId(i)->getWeaponId() != Ships[inst].weapons.secondary_bank_weapons[i]) {
+				SecondaryBanks[0]->getByBankId(i)->setWeapon(-2);
+			}
+			if (SecondaryBanks[0]->getByBankId(i)->getAmmo() != Ships[inst].weapons.secondary_bank_ammo[i]) {
+				SecondaryBanks[0]->getByBankId(i)->setAmmo(-2);
+			}
+		}
+		ship_subsys* ssl = &Ships[inst].subsys_list;
+		ship_subsys* pss;
+		for (pss = GET_FIRST(ssl); pss != END_OF_LIST(ssl); pss = GET_NEXT(pss)) {
+			model_subsystem* psub = pss->system_info;
+			if (psub->type == SUBSYSTEM_TURRET) {
+				for (auto banks : SecondaryBanks) {
+					if (banks->getSubsys() == pss) {
+						for (int i = 0; i < MAX_SHIP_SECONDARY_BANKS; i++) {
+							if (banks->getByBankId(i)->getWeaponId() != pss->weapons.secondary_bank_weapons[i]) {
+								banks->getByBankId(i)->setWeapon(-2);
+							}
+							if (banks->getByBankId(i)->getAmmo() != pss->weapons.secondary_bank_ammo[i]) {
+								banks->getByBankId(i)->setAmmo(-2);
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+}
+void ShipWeaponsDialogModel::saveShip(int inst)
+{
+	for (auto Turret : PrimaryBanks) {
+		if (Turret->getName() == "Pilot") {
+			for (auto bank : Turret->getBanks()) {
+				if (bank->getWeaponId() != -2) {
+					Ships[inst].weapons.primary_bank_weapons[bank->getBankId()] = bank->getWeaponId();
+					Ships[inst].weapons.primary_bank_ammo[bank->getBankId()] =
+						bank->getMaxAmmo() ? fl2ir(bank->getAmmo() * 100.0f / bank->getMaxAmmo()) : 0;
+				}
+			}
+		} else {
+			ship_subsys* pss = Turret->getSubsys();
+			for (auto bank : Turret->getBanks()) {
+				if (bank->getWeaponId() != -2) {
+					pss->weapons.primary_bank_weapons[bank->getBankId()] = bank->getWeaponId();
+					pss->weapons.primary_bank_ammo[bank->getBankId()] =
+						bank->getMaxAmmo() ? fl2ir(bank->getAmmo() * 100.0f / bank->getMaxAmmo()) : 0;
+				}
+			}
+		}
+	}
+	for (auto Turret : SecondaryBanks) {
+		if (Turret->getName() == "Pilot") {
+			for (auto bank : Turret->getBanks()) {
+				if (bank->getWeaponId() != -2) {
+					Ships[inst].weapons.secondary_bank_weapons[bank->getBankId()] = bank->getWeaponId();
+					Ships[inst].weapons.secondary_bank_ammo[bank->getBankId()] =
+						bank->getMaxAmmo() ? fl2ir(bank->getAmmo() * 100.0f / bank->getMaxAmmo()) : 0;
+				}
+			}
+		} else {
+			ship_subsys* pss = Turret->getSubsys();
+			for (auto bank : Turret->getBanks()) {
+				if (bank->getWeaponId() != -2) {
+					pss->weapons.secondary_bank_weapons[bank->getBankId()] = bank->getWeaponId();
+					pss->weapons.secondary_bank_ammo[bank->getBankId()] =
+						bank->getMaxAmmo() ? fl2ir(bank->getAmmo() * 100.0f / bank->getMaxAmmo()) : 0;
+				}
+			}
+		}
+	}
+}
+bool ShipWeaponsDialogModel::apply()
+{
+	if (m_isMultiEdit) {
+		object* ptr;
+		int inst;
+		ptr = GET_FIRST(&obj_used_list);
+		while (ptr != END_OF_LIST(&obj_used_list)) {
```

Comment:
Another possible infinite loop and vars that can be simplified or contained in a tighter scope.